### PR TITLE
#43 inject products details from api

### PR DESCRIPTION
**Pull request by MinhazH99**

### CODE REVIEW

File: frontend/script.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+function hideSideBar() {
+    const sidebar = document.querySelector('.hamburger');
+    sidebar.style.display = 'none';
+}
+
+// Specify the API endpoint for user data
+const apiUrl = 'http://localhost:8080/products/';
+// Make a GET request using the Fetch API
+fetch(apiUrl)
```
Comment (saraingmar) [inline frontend/script.js:14]: Just adding a note for what we talked about yesterday - I would probably wrap the `fetch` invocation in a separate helper, and place that in its own file. 

That way we could reuse the helper any time we'd want to make a request, without having to repeat the `.then`s, `if (!response.ok)` etc. (I know it's only in one place at the moment, but for the exercise of it)

File: frontend/script.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    .catch((error) => {
+        console.error('Error:', error);
+
+        for (let i = 0; i < 4; i++) {
+            if ('content' in document.createElement('template')) {
+                const trendingBody = document.querySelector('#trending');
+                const template = document.querySelector('#trending-product');
+                const clone = template.content.cloneNode(true);
+                let tp = clone.querySelectorAll('div');
+                tp[0].querySelector('img').src = './assets/images/trending-product.webp';
```
Comment (saraingmar) [inline frontend/script.js:51]: We need an alt text for the images as well.

File: frontend/script.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    })
+    .catch((error) => {
+        console.error('Error:', error);
+
+        for (let i = 0; i < 4; i++) {
+            if ('content' in document.createElement('template')) {
+                const trendingBody = document.querySelector('#trending');
+                const template = document.querySelector('#trending-product');
+                const clone = template.content.cloneNode(true);
+                let tp = clone.querySelectorAll('div');
```
Comment (saraingmar) [inline frontend/script.js:50]: What does `tp` mean in this context? Would it be possible to have a more descriptive name?

File: frontend/script.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                tp[0].textContent = 'Placeholder';
+                tp[1].textContent = '£XX.XX';
+            }
+        });
+    })
+    .catch((error) => {
+        console.error('Error:', error);
+
+        for (let i = 0; i < 4; i++) {
+            if ('content' in document.createElement('template')) {
```
Comment (saraingmar) [inline frontend/script.js:46]: Can this whole expression be separated to a variable or function, explaining what it does?

File: frontend/script.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            } else {
+                tp[0].textContent = 'Placeholder';
+                tp[1].textContent = '£XX.XX';
+            }
+        });
+    })
+    .catch((error) => {
+        console.error('Error:', error);
+
+        for (let i = 0; i < 4; i++) {
```
Comment (saraingmar) [inline frontend/script.js:45]: What does 4 mean here? Could it be a constant with a descriptive name?

File: frontend/script.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+fetch(apiUrl)
+    .then((response) => {
+        if (!response.ok) {
+            throw new Error('Network response was not ok');
+        }
+        return response.json();
+    })
+    .then((data) => {
+        data.data.forEach((element) => {
+            if ('content' in document.createElement('template')) {
```
Comment (saraingmar) [inline frontend/script.js:23]: Similar as [mentioned here](https://github.com/MinhazH99/orders101/pull/53/files#r1487585811), could we use a variable or function with a descriptive name?

File: frontend/script.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                const trendingBody = document.querySelector('#trending');
```
Comment (saraingmar) [inline frontend/script.js:24]: I would separate out all the logic which is currently placed in this callback into multiple helpers, which could then be called in a wrapping function, just to understand what everything does. 

For example 
`function createProductImage()`, `function createProductCard()`, `function formatPrice()` or similar (naming is always tricky 😄)

File: frontend/script.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                const template = document.querySelector('#trending-product');
+                const clone = template.content.cloneNode(true);
+                let tp = clone.querySelectorAll('div');
+                const img = tp[0].querySelector('img');
+                img.addEventListener('error', function (event) {
+                    event.target.src = './assets/images/trending-product.webp';
+                    event.onerror = null;
+                });
+                tp[1].textContent = element.name;
+                tp[2].textContent = '£' + element.unitPrice;
```
Comment (saraingmar) [inline frontend/script.js:34]: Can we format the price with `toLocaleString` instead of hard coding the dollar sign? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString

File: frontend/script.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                trendingBody.appendChild(clone);
+            } else {
+                tp[0].textContent = 'Placeholder';
```
Comment (saraingmar) [inline frontend/script.js:37]: Can all static texts be defined as constants, possibly placed in a separate file? (`applicationTexts` or similar?)

File: frontend/script.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+function showSideBar() {
```
Comment (saraingmar) [inline frontend/script.js:1]: Similar to how it was done for the css files (adding the `orders101-ui.css` "wrapper"), could we name this file `main.js` or similar, separate the content which is currently in it, and then import them into this file as a wrapper?

File: frontend/script.js (hunk 1 above)
Comment (saraingmar) [inline frontend/script.js:14]: Was just thinking about the immediate involation of `fetch` here. The way I've always done it is to hook up to the window load event, and only execute my script on that event - I guess in this case the page will be ready when `fetch` is called though, unless you've noticed any timing issues?

File: frontend/script.js (hunk 9 above)
Comment (josephh) [inline frontend/script.js:37]: i think `tp` is declared inside the if statement that checks templates are supported.  Then you are using `tp` here in the else block.  You might get away with this syntactically - thanks to javascript being lenient where you've missed out the 'var' or 'let' keyword but i still don't think you will get the results you want (if you get here you won't have executed the element selector...)

File: frontend/script.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        data.data.forEach((element) => {
+            if ('content' in document.createElement('template')) {
+                const trendingBody = document.querySelector('#trending');
+                const template = document.querySelector('#trending-product');
+                const clone = template.content.cloneNode(true);
+                let tp = clone.querySelectorAll('div');
+                const img = tp[0].querySelector('img');
+                img.addEventListener('error', function (event) {
+                    event.target.src = './assets/images/trending-product.webp';
+                    event.onerror = null;
```
Comment (josephh) [inline frontend/script.js:31]: what's happening here and why is this needed? 😕

Comment (MinhazH99) [reply frontend/script.js:31]: Hi Joe,

This code snippet handles scenarios where an image fails to load on a webpage. It begins by selecting the first image element (img) found within the div (tp[0])

The addEventListener function is then used to listen for the error event on the selected image. If, for any reason, the image fails to load from the server or encounters an error during loading, this event will be triggered.

In response to the error event, a callback function is invoked. This function receives the event object as its argument. Inside the callback, it's ensured that the onerror event is nullified to prevent infinite loops in case of repeated errors.

The key action takes place within the callback: the src attribute of the problematic image (event.target.src) is reassigned to a default image link ('./assets/images/trending-product.webp'). This ensures that if the original image fails to load, a predetermined fallback image is displayed instead.

File: frontend/script.js (hunk 1 above)
Comment (MinhazH99) [reply frontend/script.js]: fetch now under a separate file. Fetch function is also a function under the Product object

File: frontend/script.js (hunk 2 above)
Comment (MinhazH99) [reply frontend/script.js]: alt text added to trending product images

File: frontend/script.js (hunk 8 above)
Comment (MinhazH99) [reply frontend/script.js]: toLocaleString now used when adding the unit price for the trending product

File: frontend/script.js (hunk 5 above)
Comment (MinhazH99) [reply frontend/script.js]: 4 changed to variable named "fourPlaceholderProducts"

File: frontend/script.js (hunk 3 above)
Comment (MinhazH99) [reply frontend/script.js]: variables have been renamed to be more descriptive

File: frontend/script.js (hunk 4 above)
Comment (MinhazH99) [reply frontend/script.js]: Code has been moved in to a function called handleError which sits within the Product object

File: frontend/script.js (hunk 11 above)
Comment (MinhazH99) [reply frontend/script.js]: Looking back at this, I may have overcomplicated this. I've updated the handleError function which just reassigns the source of the image to a default image.

File: frontend/script.js (hunk 9 above)
Comment (MinhazH99) [reply frontend/script.js]: Not sure if needed anymore as static text now within index.html

Comment (MinhazH99) [reply frontend/script.js]: Ah yep. I've updated code above based off your PR request which should have solved this issue

File: frontend/script.js (hunk 7 above)
Comment (MinhazH99) [reply frontend/script.js]: Updated fetch.js to include helper functions

**Comment by saraingmar** [comment]
Should we add null checks for when selecting and modifying elements?

File: frontend/javascript/fetch.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+function fetchData(url, callback) {
```
Comment (saraingmar) [inline frontend/javascript/fetch.js:1]: We should be able to remove the `callback` argument since it's not used?

File: frontend/javascript/products.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+function createProductImage(template) {
+    let prodImage = template.querySelector('#trending-div-product-image');
+    // Need to test for null image once image is added to API
+    prodImage.setAttribute('src', './assets/images/trending-product.webp');
+    prodImage.setAttribute('alt', 'Image of a trending product');
+}
+
+function createProductLabel(template, product) {
+    let labelDiv = template.querySelector('#trending-div-product-label');
+    if (product.name == null) {
```
Comment (saraingmar) [inline frontend/javascript/products.js:31]: `name` and `unitPrice` may be `null`, but if they don't exist as properties in the response at all they will be `undefined`. One alternative is to check `if (!product.name)` instead. 

An edge case though is when the name is an empty string, or the price is 0 - in that case the "falsy" condition will be fulfilled as well, so perhaps we we could do `if (typeof test === 'undefined' || test === null)`?

File: frontend/javascript/products.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+function createProductImage(template) {
+    let prodImage = template.querySelector('#trending-div-product-image');
+    // Need to test for null image once image is added to API
+    prodImage.setAttribute('src', './assets/images/trending-product.webp');
+    prodImage.setAttribute('alt', 'Image of a trending product');
+}
+
+function createProductLabel(template, product) {
+    let labelDiv = template.querySelector('#trending-div-product-label');
```
Comment (saraingmar) [inline frontend/javascript/products.js:30]: Check if the element doesn't exist.
